packages/std/src/math.rs

```diff
@@ -106,39 +132,54 @@ impl ops::Sub for Uint128 {
     }
 }
 
-impl ops::Mul<Decimal9> for Uint128 {
+/// Both d*u and u*d with d: Decimal and u: Uint128 returns an Uint128. There is no
+/// specific reason for this decision other than the initial use cases we have. If you
+/// need a Decimal result for the same calculation, use Decimal(d*u) or Decimal(u*d).
+impl ops::Mul<Decimal> for Uint128 {
     type Output = Self;
 
-    fn mul(self, rhs: Decimal9) -> Self {
-        self.multiply_ratio(rhs.0.into(), DECIMAL_FRACTIONAL.into())
+    #[allow(clippy::suspicious_arithmetic_impl)]
+    fn mul(self, rhs: Decimal) -> Self::Output {
+        // 0*a and b*0 is always 0
+        if self.is_zero() || rhs.is_zero() {
+            return Uint128::zero();
+        }
+        self.multiply_ratio(rhs.0, DECIMAL_FRACTIONAL)
+    }
+}
+
+impl ops::Mul<Uint128> for Decimal {
+    type Output = Uint128;
+
+    fn mul(self, rhs: Uint128) -> Self::Output {
+        rhs * self
     }
 }
 
 impl Uint128 {
     /// returns self * num / denom
     pub fn multiply_ratio(&self, num: Uint128, denom: Uint128) -> Uint128 {
-        // special case for 0/0 (== 1)
-        if num.0 == 0 && denom.0 == 0 {
-            return *self;
+        if denom.0 == 0 {
+            panic!("Denominator must not be zero");
         }
         // TODO: minimize rounding that takes place (using gcd algorithm)
         let val = self.u128() * num.u128() / denom.u128();
         Uint128::from(val)
     }
 
-    /// Returns the ratio (self / denom) as Decimal9 fixed-point
-    pub fn calc_ratio(&self, denom: Uint128) -> Decimal9 {
+    /// Returns the ratio (self / denom) as Decimal fixed-point
+    pub fn calc_ratio(&self, denom: Uint128) -> Decimal {
         // special case: 0/0 = 1.0
         if self.0 == 0 && denom.0 == 0 {
-            return Decimal9::one();
+            return Decimal::one();
```

Comment:
You wanted to remove this, right? I agree to move this to the contract

Reply:
Oh good point. I did not look into this function, just changed it in multiply_ratio.

Reply:
Done in https://github.com/CosmWasm/cosmwasm/pull/327/commits/73e165f16d4da676088e794190699ff015711589